### PR TITLE
loading water dialer async

### PR DESCRIPTION
**Pull request by WendelHime**
This pull request requires #1425 and #1427 to be merged.

This pull request goal is to load the WATER dialer asynchronously so when flashlight start initializing all proxies that it could find it doesn't keep waiting until WATER dialer is ready (it can take some time since the WASM need to be downloaded)

### CODE REVIEW

File: chained/water_impl.go
Comment (garmr-ulfr) [inline chained/water_impl.go]: Check if what is the intended behavior? There are places where `d.dialer` can be `nil`, so unless I'm missing something, this is correct. Although, we should wrap the error that occurred while trying to download the WASM. That way, we would have a bit more info on why it failed in the traces instead of only in the logs. 
Maybe something like:
```
d.wgDownload.Wait()
if d.dialer == nil {
	d.downloadErrMx.Lock()
	err := d.downloadErr
	d.downloadErrMx.Unlock()

	return nil, log.Errorf("dialer not available: %w", err)
}
```

File: chained/water_impl.go
Comment (garmr-ulfr) [inline chained/water_impl.go:63]: Nit: Why a `WaitGroup`? Since it's only one goroutine we'd be waiting on, we should probably use a `chan struct{}` to signal it's done. I know it's a small thing but, it would help with readability and maintainability so the reader knows it's only one goroutine.

File: chained/water_impl.go
Comment (garmr-ulfr) [inline chained/water_impl.go]: Nit: Isn't this basically `else if wasmAvailableAt != ""`?

**Review by garmr-ulfr — COMMENTED** [review]
Just a couple of things.

File: chained/water_impl.go
Comment (WendelHime) [reply chained/water_impl.go:63]: I've chosen to use a wait group because I can use it at tests to know when it loaded ~~correctly~~ the WASM file (when downloading). But it also makes sense to use a channel

File: chained/water_impl.go
Comment (WendelHime) [reply chained/water_impl.go]: I'll follow the suggestion for adding the error, but I don't believe we need the mutex in this case, since the waitgroup is already saying the operation is done, the error wouldn't be accessed or modified anywhere else.

Comment (garmr-ulfr) [reply chained/water_impl.go]: > I don't believe we need the mutex in this case, since the waitgroup is already saying the operation is done

Oh, right. That's a good point.

File: chained/water_impl.go
Comment (garmr-ulfr) [reply chained/water_impl.go:63]: Oh OK, that makes sense. This is fine as is and it works, so it's not an issue!

**Review by garmr-ulfr — APPROVED** [review]
LGTM

File: chained/water_impl.go
Comment (hwh33) [inline chained/water_impl.go:120]: Hey @WendelHime sorry I just saw this logic.  Won't this lead to the client still getting blocked until the WASM download finishes?  The client will try to use this dialer, but the call to `waterImpl.dialServer` will block on this line.  I think you need to respect the input context, like:
```
select {
case <-someChannelWaitingOnWaterDownload:
    // carry on with dial
case <-ctx.Done():
    return nil, log.Errorf("context completed while waiting for WASM download: %w", ctx.Error())
}
```

However, even that has issues as the WATER dialer will then appear to be non-functional.  In the current code, the bandit dialer will simply [ignore it](https://github.com/getlantern/flashlight/blob/7fbdf0b33ce9bf0a8d13a7c66a90ab1d06b0b01d/bandit/bandit.go#L208-L213).  Even if that code changes, errors returned by a proxy dialer are always going to lead to that dialer being de-prioritized.  We don't want the WATER dialer to always get ignored or de-prioritized.

Instead, we need the WATER dialer to be added to the overall set of dialers only when it's ready for use.  I realize that is a more involved change.  Sorry that this comment is coming after this work was already reviewed and merged.

Comment (WendelHime) [reply chained/water_impl.go:120]: 🤔  can we introduce a new dialer function flag that says the dialer is ready? I'm considering these steps:

1. Create a `IsReady() bool` function at [bandit dialer interface](https://github.com/getlantern/flashlight/blob/main/bandit/bandit.go#L275).
2. Implement `IsReady` to all chained implementations.
3. Update water impl dialer to update the status properly.
4. Update [bandit logic for checking](https://github.com/getlantern/flashlight/blob/main/bandit/bandit.go#L207), if the dialer is not ready continue search of dialers.

This solve the issue for bandit, but not for bypass

Comment (hwh33) [reply chained/water_impl.go:120]: I like that approach!  Looks like bypass will get access to that function as well since `chained.CreateDialersMap` returns a map of `bandit.Dialers`.

Comment (garmr-ulfr) [reply chained/water_impl.go:120]: > In the current code, the bandit dialer will simply [ignore it](https://github.com/getlantern/flashlight/blob/7fbdf0b33ce9bf0a8d13a7c66a90ab1d06b0b01d/bandit/bandit.go#L208-L213). Even if that code changes, errors returned by a proxy dialer are always going to lead to that dialer being de-prioritized. We don't want the WATER dialer to always get ignored or de-prioritized.

Sorry, this had not occurred to me. Good catch!

Comment (WendelHime) [reply chained/water_impl.go:120]: PR ready for review at: https://github.com/getlantern/flashlight/pull/1436

**Pull request merged**
